src/openrct2/Context.cpp

```diff
@@ -655,13 +655,24 @@ namespace OpenRCT2
             ContextOpenIntent(&intent);
         }
 
-        void SetProgress(uint32_t currentProgress, uint32_t totalCount, StringId format = STR_NONE) override
+        void SetProgress(
+            uint32_t currentProgress, uint32_t totalCount, StringId format = STR_NONE, bool forceDraw = false) override
         {
             auto intent = Intent(INTENT_ACTION_PROGRESS_SET);
             intent.PutExtra(INTENT_EXTRA_PROGRESS_OFFSET, currentProgress);
             intent.PutExtra(INTENT_EXTRA_PROGRESS_TOTAL, totalCount);
             intent.PutExtra(INTENT_EXTRA_STRING_ID, format);
             ContextOpenIntent(&intent);
+
+            // Ideally, we'd force a redraw at all times at this point. OpenGL has to be directed
+            // from the main thread, though, so this cannot be invoked when off main thread.
+            // It's fine (and indeed useful!) for synchronous calls, so we keep it as an option.
+            if (!gOpenRCT2Headless && forceDraw)
+            {
+                _uiContext->ProcessMessages();
+                WindowUpdateAll();
+                Draw();
+            }
```

Comment:
Ideally these 'force redraws' only redraw the portion of the screen concerned with the progress window. Currently, it redraws the viewport as well, which can be a bit glitchy, considering we're messing with the (global!) game state while we're loading the next file. I'd like to investigate if/how we can potentially do this before putting this forward for review. Suggestions are welcome.